### PR TITLE
Feature/more abstract methods

### LINKED ISSUE #98: Shape2D Abstract Properties (perimeter, planar moments of inertia)
### Feature description

I noticed that there are some abstract properties of `Shape3D` that are not mirrored for `Shape2D`.

`Shape3D` defines abstract volume, surface area, and inertia tensor properties.

`Shape2D` defines only area.

### Proposed solution

I suggest adding perimeter and planar moments of inertia to the `Shape2D` abstract properties.

I believe the perimeter property is not yet implemented for polygon or its convex/sphero variants.

### PR DESCRIPTION
**Pull request by vyasr**
## Description

Expands the minimum expected API for anything to be a shape. In the process defined perimeter calculations for polygons and spheropolygons, and got a few other properties for free across the board by moving general methods up the hierarchy (for instance iq).

## Motivation and Context

Resolves #98 and #91. I exercised some judgment in which properties _have_ to be implemented (i.e. which are abstract) and which ones _should_ be implemented (concrete methods/properties that raise `NotImplementedError`). I tried to be pragmatic with this decision more than anything else. I also removed the abstractmethod setters since they don't do anything; we'll rely on the tests @Tobias-Dwyer is writing to manually verify these.

## Types of Changes

- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/coxeter/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/coxeter/blob/master/ContributorAgreement.md).
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.
- [x] I have updated the [changelog](https://github.com/glotzerlab/coxeter/blob/master/ChangeLog.txt) and the [credits](https://github.com/glotzerlab/coxeter/blob/master/doc/source/credits.rst).

### CODE REVIEW

File: ChangeLog.rst [hunk 1]
```diff
@@ -10,6 +10,7 @@ Added
 
 -  Circumsphere and insphere from center calculations for ConvexSpheropolyhedron.
 -  Shape families associated with a DOI can be directly accessed via a dictionary.
+-  Expected abstract interface for shapes (both 2D and 3D) has expanded.
```
Comment (bdice) [inline ChangeLog.rst:13]: This changelog format is a little weird, we should not need two spaces after the dash. I notice there's inconsistency with the "Fixed" section below.

File: coxeter/shape_classes/base_classes.py [hunk 2]
```diff
@@ -18,15 +20,27 @@ def center(self):
...
-    @center.setter
-    @abstractmethod
-    def center(self, value):
-        pass
-
     @property
     def gsd_shape_spec(self):
         """dict: Get a :ref:`complete GSD specification <shapes>`."""  # noqa: D401
-        return {}
+        raise NotImplementedError
```
Comment (bdice) [inline coxeter/shape_classes/base_classes.py:26]: Good, I was going to ask about this.

File: coxeter/shape_classes/base_classes.py [hunk 3]
```diff
@@ -38,11 +52,45 @@ def area(self):
...
+        r"""list[float, float, float]: Get the planar and product moments of inertia.
+
+        Moments are computed with respect to the :math:`x` and :math:`y`
+        axes. In addition to the two planar moments, this property also
+        provides the product of inertia.
+
+        The `planar moments <https://en.wikipedia.org/wiki/Polar_moment_of_inertia>`__
+        and the
+        `product <https://en.wikipedia.org/wiki/Second_moment_of_area#Product_moment_of_area>`__
+        of inertia define the in-plane area distribution.
```
Comment (bdice) [inline coxeter/shape_classes/base_classes.py:72]: I would prefer to give the full term in the link title: `planar moments of inertia <...>` and `product moment of area <...>`.

File: coxeter/shape_classes/base_classes.py [hunk 4]
```diff
@@ -38,11 +52,45 @@ def area(self):
...
+        normal vector).
+
+        The polar moment is computed as the sum of the two planar moments of inertia.
+        """  # noqa: E501
+        return np.sum(self.planar_moments_inertia[:2])
+
+    @property
+    def iq(self):
+        """float: The isoperimetric quotient."""
+        # TODO: Add reference to math and some resource in the docstring.
```
Comment (bdice) [inline coxeter/shape_classes/base_classes.py:91]: I recommend we solve this TODO now. Here's two links I like:
1. https://mathworld.wolfram.com/IsoperimetricQuotient.html
2. https://en.wikipedia.org/wiki/Isoperimetric_inequality#On_a_plane

File: coxeter/shape_classes/base_classes.py [hunk 5]
```diff
@@ -53,23 +101,18 @@ def volume(self):
...
-    @abstractmethod
     def inertia_tensor(self):
         """:math:`(3, 3)` :class:`numpy.ndarray`: Get the inertia tensor."""
-        pass
+        raise NotImplementedError
+
+    @property
+    def iq(self):
+        """float: The isoperimetric quotient."""
+        # TODO: allow for non-spherical reference ratio (changes the prefactor)
```
Comment (bdice) [inline coxeter/shape_classes/base_classes.py:117]: Isn't using a non-spherical reference the same as taking the ratio of two IQs that both use the spherical reference? That is, isn't this a trivial thing to ask users to factor out?

(Reminds me of the change of base formula, `log10(x) = ln(x)/ln(10)`, for some reason.)

File: tests/test_polygon.py [hunk 6]
```diff
@@ -313,3 +313,22 @@ def test_incircle_from_center(convex_square):
     circle = convex_square.incircle_from_center
     assert np.all(circle.center == convex_square.center)
     assert circle.radius == 0.5
+
+
+@pytest.mark.parametrize("num_sides", range(3, 6))
+def test_perimeter(num_sides):
+    """Test the polygon perimeter calculation."""
+
+    def unit_area_regular_ngon_sidelength(n):
```
Comment (bdice) [inline tests/test_polygon.py:322]:
```suggestion
    def unit_area_regular_n_gon_side_length(n):
```

File: tests/test_polygon.py [hunk 7]
```diff
@@ -313,3 +313,22 @@ def test_incircle_from_center(convex_square):
...
+        r"""Compute the side length of a unit-area regular polygon analytically.
+
+        The area of regular n-gon is given by
+        :math:`\\frac{s^2 n}{4 \\tan\\left(\\frac{180}{n}\\right)}`. This function sets
```
Comment (bdice) [inline tests/test_polygon.py:326]: raw string, too many slashies ⚔️
```suggestion
        :math:`\frac{s^2 n}{4 \tan\left(\frac{180}{n}\right)}`. This function sets
```

**Review by bdice — CHANGES_REQUESTED** [review]

File: ChangeLog.rst (hunk 1 above)
Comment (vyasr) [reply ChangeLog.rst:13]: I agree that it's strange, tbh I assumed you had changed it to this formatting at some point and didn't feel like doing anything about it. We can revise it later.

File: coxeter/shape_classes/base_classes.py (hunk 5 above)
Comment (vyasr) [reply coxeter/shape_classes/base_classes.py:117]: Yes, it is. I think someone (maybe @tcmoore3) made this request, which is how the TODO got added.

File: ChangeLog.rst (hunk 1 above)
Comment (vyasr) [reply ChangeLog.rst:13]: #108

**Review by bdice — APPROVED** [review]
LGTM!

**Pull request merged**